File: sample/src/main/java/me/xiaopan/spear/sample/MyApplication.java
/*
 * Copyright (C) 2013 Peng fei Pan <sky@xiaopan.me>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.spear.sample;

import android.app.Application;
import android.graphics.Color;
import android.text.format.Formatter;
import android.util.Log;
import android.widget.ImageView;

import me.xiaopan.android.gohttp.GoHttp;
import me.xiaopan.spear.DisplayOptions;
import me.xiaopan.spear.LoadOptions;
import me.xiaopan.spear.Spear;
import me.xiaopan.spear.display.ColorTransitionImageDisplayer;
import me.xiaopan.spear.display.TransitionImageDisplayer;
import me.xiaopan.spear.process.CircleImageProcessor;
import me.xiaopan.spear.process.GaussianBlurImageProcessor;
import me.xiaopan.spear.sample.util.Settings;

public class MyApplication extends Application {

    @Override
    public void onCreate() {
        super.onCreate();

        GoHttp.with(getBaseContext()).setDebugMode(true);
        Spear.setDebugMode(true);

        Spear.putOptions(OptionsType.Rectangle,
                new DisplayOptions(getBaseContext())
                        .setLoadingImage(R.drawable.image_loading2)
                        .setFailureImage(R.drawable.image_load_fail2)
                        .setPauseDownloadImage(R.drawable.image_click2)
                        .setDecodeGifImage(false)
                        .setImageDisplayer(new TransitionImageDisplayer())
        );

        Spear.putOptions(
                OptionsType.Detail,
                new DisplayOptions(getBaseContext())
                        .setFailureImage(R.drawable.image_load_fail2)
                        .setPauseDownloadImage(R.drawable.image_click2)
                        .setImageDisplayer(new ColorTransitionImageDisplayer(Color.TRANSPARENT))
        );

        Spear.putOptions(
                OptionsType.Circular,
                new DisplayOptions(getBaseContext())
                        .setLoadingImage(R.drawable.image_loading2, CircleImageProcessor.getInstance())
                        .setFailureImage(R.drawable.image_load_fail2, CircleImageProcessor.getInstance())
                        .setPauseDownloadImage(R.drawable.image_click2, CircleImageProcessor.getInstance())
                        .setDecodeGifImage(false)
                        .setImageDisplayer(new TransitionImageDisplayer())
                        .setImageProcessor(CircleImageProcessor.getInstance())
        );

        Spear.putOptions(
                OptionsType.WindowBackground,
                new LoadOptions(getBaseContext())
                        .setScaleType(ImageView.ScaleType.CENTER_CROP)
                        .setImageProcessor(new GaussianBlurImageProcessor(true))
                        .setDecodeGifImage(false)
        );

        boolean isPauseDownload = Settings.with(getBaseContext()).isMobileNetworkPauseDownload();
        Spear.with(getBaseContext()).getConfiguration().setMobileNetworkPauseDownload(isPauseDownload);
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();

        Log.w("Application", "Memory is very low, has automatic releasing Spear in memory cache(" + Formatter.formatFileSize(getBaseContext(), Spear.with(getBaseContext()).getConfiguration().getMemoryCache().getSize()) + ")");
        Spear.with(getBaseContext()).getConfiguration().getMemoryCache().clear();
    }
}

File: sample-compose/src/main/java/com/github/panpf/sketch/compose/sample/MyApplication.kt
/*
 * Copyright (C) 2019 panpf <panpfpanpf@outlook.com>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.panpf.sketch.compose.sample

import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import androidx.multidex.MultiDexApplication
import com.github.panpf.sketch.Sketch.Builder
import com.github.panpf.sketch.SketchConfigurator
import com.github.panpf.sketch.decode.ApkIconBitmapDecoder
import com.github.panpf.sketch.decode.AppIconBitmapDecoder
import com.github.panpf.sketch.decode.FFmpegVideoFrameBitmapDecoder
import com.github.panpf.sketch.decode.GifAnimatedDrawableDecoder
import com.github.panpf.sketch.decode.GifDrawableDrawableDecoder
import com.github.panpf.sketch.decode.GifMovieDrawableDecoder
import com.github.panpf.sketch.decode.HeifAnimatedDrawableDecoder
import com.github.panpf.sketch.decode.SvgBitmapDecoder
import com.github.panpf.sketch.decode.VideoFrameBitmapDecoder
import com.github.panpf.sketch.decode.WebpAnimatedDrawableDecoder
import com.github.panpf.sketch.fetch.AppIconUriFetcher
import com.github.panpf.sketch.http.OkHttpStack
import com.github.panpf.sketch.request.PauseLoadWhenScrollingDisplayInterceptor
import com.github.panpf.sketch.request.SaveCellularTrafficDisplayInterceptor
import com.github.panpf.sketch.util.Logger
import com.github.panpf.sketch.util.Logger.Level.DEBUG

class MyApplication : MultiDexApplication(), SketchConfigurator {

    override fun createSketchConfig(): Builder.() -> Unit = {
        logger(Logger(DEBUG))
        httpStack(OkHttpStack.Builder().build())
        addRequestInterceptor(SaveCellularTrafficDisplayInterceptor())
        addRequestInterceptor(PauseLoadWhenScrollingDisplayInterceptor())
        components {
            addFetcher(AppIconUriFetcher.Factory())

            addBitmapDecoder(SvgBitmapDecoder.Factory())
            addBitmapDecoder(ApkIconBitmapDecoder.Factory())
            addBitmapDecoder(AppIconBitmapDecoder.Factory())
            addBitmapDecoder(
                if (VERSION.SDK_INT >= VERSION_CODES.O_MR1) {
                    VideoFrameBitmapDecoder.Factory()
                } else {
                    FFmpegVideoFrameBitmapDecoder.Factory()
                }
            )

            addDrawableDecoder(
                when {
                    VERSION.SDK_INT >= VERSION_CODES.P -> GifAnimatedDrawableDecoder.Factory()
                    VERSION.SDK_INT >= VERSION_CODES.KITKAT -> GifMovieDrawableDecoder.Factory()
                    else -> GifDrawableDrawableDecoder.Factory()
                }
            )
            if (VERSION.SDK_INT >= VERSION_CODES.P) {
                addDrawableDecoder(WebpAnimatedDrawableDecoder.Factory())
            }
            if (VERSION.SDK_INT >= VERSION_CODES.P) {
                addDrawableDecoder(HeifAnimatedDrawableDecoder.Factory())
            }
        }
    }
}